# 7장 오류 처리

> 📅 작성일: 2025-04-16

## 📌 핵심 요약
오류 처리는 중요하지만, 오류 처리 코드를 여기 저기 흩어지게 만들어 프로그램 논리를 이해하기 어렵게 하면, 깨끗한 코드라고 부르기 힘들다.

## 오류 코드보다 예외(Exceptions)를 사용하라

- 예전 프로그래밍 언어들은 exception을 제공하지 않았음.

  ☠️ 나쁜 예시

    ```java
    // Bad
    public class DeviceController {
      ...
      public void sendShutDown() {
        DeviceHandle handle = getHandle(DEV1);
        // Check the state of the device
        if (handle != DeviceHandle.INVALID) {
          // Save the device status to the record field
          retrieveDeviceRecord(handle);
          // If not suspended, shut down
          if (record.getStatus() != DEVICE_SUSPENDED) {
            pauseDevice(handle);
            clearDeviceWorkQueue(handle);
            closeDevice(handle);
          } else {
            logger.log("Device suspended. Unable to shut down");
          }
        } else {
          logger.log("Invalid handle for: " + DEV1.toString());
        }
      }
      ...
    }
    ```

    - 에러 flag를 set하거나 에러 코드를 리턴했었음
        - ⇒ 예외 처리를 잊어비리기 쉽고, 로직을 헷갈리게 하기 쉬움.
        - ⇒ exceptions를 사용하자

  👼 좋은 예시

    ```java
    // Good
    public class DeviceController {
      ...
      public void sendShutDown() {
        try {
          tryToShutDown();
        } catch (DeviceShutDownError e) {
          logger.log(e);
        }
      }
        
      private void tryToShutDown() throws DeviceShutDownError {
        DeviceHandle handle = getHandle(DEV1);
        DeviceRecord record = retrieveDeviceRecord(handle);
        pauseDevice(handle); 
        clearDeviceWorkQueue(handle); 
        closeDevice(handle);
      }
      
      private DeviceHandle getHandle(DeviceID id) {
        ...
        throw new DeviceShutDownError("Invalid handle for: " + id.toString());
        ...
      }
      ...
    }
    ```


## Try-Catch-Finally 문부터 작성해라

- try 문은 transaction 처럼 동작하는 실행코드로, catch 문은 try 문에 관계없이 프로그램을 일관적인 상태로 유지하도록 함
- 이렇게 함으로써 코드의 Scope 정의가 가능해진다.

예시 - 파일이 없으면 예외를 던지는지 알아보는 단위테스트

```java
@Test(expected = StorageException.class)
public void retrieveSectionShouldThrowOnInvalidFileName() {
	sectionStore.retreiveSection("invalid-file")
}	
```

예시 - 위 단위 테스트에 맞춰 구현한 코드

```java
    // Step 1: StorageException을 던지지 않으므로 이 테스트는 실패한다.
  public List<RecordedGrip> retrieveSection(String sectionName) {
    // dummy return until we have a real implementation
    return new ArrayList<RecordedGrip>();
  }
  
    // Step 2: 이제 테스트는 통과한다.
  public List<RecordedGrip> retrieveSection(String sectionName) {
    try {
      FileInputStream stream = new FileInputStream(sectionName)
    } catch (Exception e) {
      throw new StorageException("retrieval error", e);
    }
  return new ArrayList<RecordedGrip>();
}

  // Step 3: Exception의 범위를 FileNotFoundException으로 줄여 정확히
   // 어떤 Exception이 발생한지 체크하자.
   
  public List<RecordedGrip> retrieveSection(String sectionName) {
    try {
      FileInputStream stream = new FileInputStream(sectionName);
      stream.close();
    } catch (FileNotFoundException e) {
      throw new StorageException("retrieval error", e);
    }
    return new ArrayList<RecordedGrip>();
  }
```

## Unchecked Excpetions를 사용해라

- 안정된 소프트웨어를 제작하는 요소로 확인된 예외가 반드시 필요하진 않음
- **checked excpetion은 OCP를 위반함.**
    - 하위 단계에서 코드를 변경하면 상위 단계 메서드 선언부를 모두 고쳐야 함.

## Exceptions로 문맥, 의미를 제공해라

- 예외를 던질 때는 전후 상황을 충분히 덧붙여야 한다.

- 오류 메세지에 정보를 담아 예외와 함께 던져야 함.
-

<aside>
✅

오류 메시지에 정보를 담아 예외와 함께 던져야 한다.

- 실패한 연산 이름과 실패 유형도 언급
</aside>

## 호출자를 고려해 예외 클래스를 정의하라

- 오류가 발생한 위치로 오류 분류가 가능하다.
    - 예를 들어, 디바이스 실패, 프로그래밍 오류 등으로 분류
- 가장 중요한 것은, `오류를 잡아내는 방법` 이 되어야 한다.

안좋은 예시 - 외부 라이브러리를 호출하는 try-catch-finally 문을 포함한 코드

```java
  // Bad
  // catch문의 내용이 거의 같다.
  
ACMEPort port = new ACMEPort(12);
  try {
    port.open();
  } catch (DeviceResponseException e) {
    reportPortError(e);
    logger.log("Device response exception", e);
  } catch (ATM1212UnlockedException e) {
    reportPortError(e);
    logger.log("Unlock exception", e);
  } catch (GMXError e) {
    reportPortError(e);
    logger.log("Device response exception");
  } finally {
    ...
  }
```

```java
  // Good
  // ACME 클래스를 LocalPort 클래스로 래핑해 new ACMEPort().open() 메소드에서 던질 수 있는 exception들을 간략화
  
  LocalPort port = new LocalPort(12);
  try {
    port.open();
  } catch (PortDeviceFailure e) {
    reportError(e);
    logger.log(e.getMessage(), e);
  } finally {
    ...
  }
  
  // 아래 LocalPort 클래스는 ACMEPort 클래스가 던지는 예외를 잡아 반환하는 Wrapper일 뿐이다.
  public class LocalPort {
    private ACMEPort innerPort;
    public LocalPort(int portNumber) {
      innerPort = new ACMEPort(portNumber);
    }
    
    public void open() {
      try {
        innerPort.open();
      } catch (DeviceResponseException e) {
        throw new PortDeviceFailure(e);
      } catch (ATM1212UnlockedException e) {
        throw new PortDeviceFailure(e);
      } catch (GMXError e) {
        throw new PortDeviceFailure(e);
      }
    }
    ...
  }
```

- 외부 API를 감싸면 외부 라이브러리와 프로그램 사이에서 의존성이 크게 줄어듦

## 정상 흐름을 정의하라

- 일반적으로는 위에서 봤던 방식들이 유용하지만, catch 문에서 예외적인 상황(special case)을 처리해야 하는 경우 코드가 더러워지는 일이 발생할 수 있다.
-

## null을 반환하지 마라

```java
  // BAD!!!!

  public void registerItem(Item item) {
    if (item != null) {
      ItemRegistry registry = peristentStore.getItemRegistry();
      if (registry != null) {
        Item existing = registry.getItem(item.getID());
        if (existing.getBillingPeriod().hasRetailOwner()) {
          existing.register(item);
        }
      }
    }
  }
  
  
  // 위 peristentStore가 null인 경우에 대한 예외처리가 안된 것을 눈치챘는가?
  // 만약 여기서 NullPointerException이 발생했다면 수십단계 위의 메소드에서 처리해줘야 하나?
  // 이 메소드의 문제점은 null 체크가 부족한게 아니라 null체크가 너무 많다는 것이다.
```

- 메서드에서 null을 반환하고픈 유혹이 든다면 그 대신 예외를 던지거나 특수 사례 객체를 반환한다.
- 사용하려는 외부 API가 null을 반환한다면 Wrapper 메서드를 구현해 예외를 던지거나 특수 사례 객체를 반환하는 방식을 고려.
- 많은 경우, 특수 사례 객체가 손쉬운 해결책이다.

```java
  // Bad
  List<Employee> employees = getEmployees();
  if (employees != null) {
    for(Employee e : employees) {
      totalPay += e.getPay();
    }
  }
```

- getEmployees는 null도 반환한다. 하지만, 반드시 null을 반환할 필요는 없다.
    - getEmployees를 변경해 빈 리스트를 반환한다면, 코드가 훨씬 깔끔해진다.

    ```java
      // Good
      List<Employee> employees = getEmployees();
      for(Employee e : employees) {
        totalPay += e.getPay();
      }
      
      public List<Employee> getEmployees() {
        if( .. there are no employees .. )
          return Collections.emptyList();
        }
    }
    ```


## null을 전달하지 마라

- null을 리턴하는 방식도 나쁘지만, null을 메서드로 넘기는 것이 더 나쁘다.

```java
// Bad
// calculator.xProjection(null, new Point(12, 13));
// 위와 같이 부를 경우 NullPointerException 발생
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    return (p2.x – p1.x) * 1.5;
  }
  ...
}

// Bad
// NullPointerException은 안나지만 윗단계에서 InvalidArgumentException이 발생할 경우 처리해줘야 함.
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    if(p1 == null || p2 == null){
      throw InvalidArgumentException("Invalid argument for MetricsCalculator.xProjection");
    }
    return (p2.x – p1.x) * 1.5;
  }
}

// Bad
// 좋은 명세이지만 첫번째 예시와 같이 NullPointerException 문제를 해결하지 못한다.
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    assert p1 != null : "p1 should not be null";
    assert p2 != null : "p2 should not be null";
    
    return (p2.x – p1.x) * 1.5;
  }
}
```

## 💭 느낀 점

- error flag를 리턴하는 것보다는 exception으로 처리하는 것이 낫다는 것을 처음 알게 되었다. 앞으로는 해당 방법을 이용해서 에러 핸들링을 해야겠다고 생각했다.